core/trino-main/src/main/java/io/trino/server/security/oauth2/OAuth2Service.java

```diff
@@ -152,16 +152,7 @@ else if (STATE_AUDIENCE_REST.equals(stateClaims.getAudience())) {
                 .parseClaimsJws(accessToken.getAccessToken())
                 .getBody();
 
-        // user did not grant the requested scope `openid` so we don't have the ID token and therefore can't validate the nonce
-        // or the ID token is present but the nonce is not which means that someone is trying to obtain the token which he is not supposed to get
-        if (nonce.isPresent() != accessToken.getIdToken().isPresent()) {
-            throw new ChallengeFailedException("Cannot validate nonce parameter");
-        }
-        // validate nonce from the ID token
-        nonce.ifPresent(n -> Jwts.parser()
-                .setSigningKeyResolver(signingKeyResolver)
-                .require(NONCE, hashNonce(n))
-                .parseClaimsJws(accessToken.getIdToken().get()));
+        validateNonce(stateClaims, accessToken, nonce);
```

Comment:
nit: this could be two commits. One for method extraction and other for change. That way it is easier to see what actually changed.